File: base_test.go
// Copyright 2023 Felipe Zipitria
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"fmt"
	"testing"

	"github.com/antlr4-go/antlr/v4"

	"github.com/fzipi/seclang_parser/parsing"
)

type TreeShapeListener struct {
	*parsing.BaseSecLangParserListener
}

func NewTreeShapeListener() *TreeShapeListener {
	return new(TreeShapeListener)
}

func (t *TreeShapeListener) EnterEveryRule(ctx antlr.ParserRuleContext) {
	fmt.Println(ctx.GetText())
}

var testfiles = []string{
	"testdata/test1.conf",
	"testdata/test2.conf",
	"testdata/test3.conf",
	"testdata/test4.conf",
	"testdata/test5.conf",
	"testdata/crs-setup.conf",
	"testdata/REQUEST-901-INITIALIZATION.conf",
}

func TestSampleFiles(t *testing.T) {
	for _, file := range testfiles {
		input, err := antlr.NewFileStream(file)
		if err != nil {
			t.Errorf("Error reading file %s", file)
			continue
		}
		lexer := parsing.NewSecLangLexer(input)
		for {
			token := lexer.NextToken()
			if token.GetTokenType() == antlr.TokenEOF {
				break
			}
			fmt.Printf("%s (%q)\n",
				lexer.SymbolicNames[token.GetTokenType()], token.GetText())
		}
		stream := antlr.NewCommonTokenStream(lexer, 0)
		p := parsing.NewSecLangParser(stream)
		p.AddErrorListener(antlr.NewDiagnosticErrorListener(true))
		p.BuildParseTrees = true
		tree := p.Configuration()
		antlr.ParseTreeWalkerDefault.Walk(NewTreeShapeListener(), tree)
	}
}

File: base_test.py
# Copyright 2025 Felipe Zipitria
# SPDX-License-Identifier: Apache-2.0

import sys
from antlr4 import *
from parsing.SecLangLexer import SecLangLexer
from parsing.SecLangParser import SecLangParser
from parsing.SecLangParserListener import SecLangParserListener

class Rule:
    def __init__(self):
        pass

class SecLangListener(SecLangParserListener):
    def enterStmt(self, ctx):
        print("Entering statement:", ctx.getText())
        Rule(variables, operators, actions)

    def enterAction(self, ctx:SecLangParser.ActionContext):
        print("Entering action:", ctx.getText())

    def enterConfig_value_types(self, ctx:SecLangParser.Config_value_typesContext):
        print("Entering config value:", ctx.getText())


def main(argv):
    if len(sys.argv) > 1:
        input = FileStream(sys.argv[1])
    else:
        input = InputStream(sys.stdin.readline())
    lexer = SecLangLexer(input)
    stream = CommonTokenStream(lexer)
    parser = SecLangParser(stream)
    tree = parser.configuration()
    # walk and print
    printer = SecLangListener()
    walker = ParseTreeWalker()
    walker.walk(printer, tree)


if __name__ == '__main__':
    main(sys.argv)
